### PR TITLE
fix!: insert logging layer only if requested

### PR DESCRIPTION
**Pull request by coryan**
**BREAKING CHANGE**: the `google::cloud::storage::Client` constructor
now calls `RawClient::client_options()`, which may need changes in how
you initialize any Mocks for this class.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3349)

### CODE REVIEW

File: google/cloud/storage/client_write_object_test.cc [hunk 1]
```diff
@@ -149,6 +149,7 @@ TEST_F(WriteObjectTest, WriteObjectPermanentSessionFailurePropagates) {
   EXPECT_CALL(*mock, CreateResumableSession(_)).WillOnce(Invoke(returner));
   EXPECT_CALL(*mock_session, UploadChunk(_))
       .WillRepeatedly(Return(PermanentError()));
+  EXPECT_CALL(*mock_session, next_expected_byte()).WillRepeatedly(Return(0));
```
Comment (devbww) [inline google/cloud/storage/client_write_object_test.cc]: Unrelated?

Comment (coryan) [reply google/cloud/storage/client_write_object_test.cc]: Yes, will fix separately.
